src/org/rascalmpl/library/lang/paths/Unix.rsc

```diff
@@ -0,0 +1,108 @@
+}
+module lang::paths::Unix
+
+lexical UnixPath
+    = absolute: Slashes UnixFilePath?
+    | relative: UnixFilePath 
+    | home    : "~" (Slashes UnixFilePath)?
+    | user    : "~" UserName name (Slashes UnixFilePath)?
+    ;
+
+lexical UserName = ![/~]+;
+
+lexical PathChar = ![/];
+
+lexical PathSegment
+    = current: "."
+    | parent : ".."
+    | name   : (PathChar \ "~" PathChar*) \ ".." \ "." \ "~"
+    ;
+
+lexical Slashes = Slash+ !>> [/];
+
+lexical Slash = [/];
+
+lexical UnixFilePath = {PathSegment Slashes}+ segments Slashes?;
+
+import ParseTree;
+
+@synopsis{Convert a Unix path literal to a source location URI}
+@description{
+1. parses the path using the grammar for ((UnixPath))
+2. takes the _literal_ name components using string interpolation `"<segment>"`. This means no decoding/encoding happens at all while extracting
+hostname, share name and path segment names. Also all superfluous path separators are skipped.
+3. uses `loc + str` path concatenation with its builtin character encoding to construct the URI. Also
+the right path separators are introduced. 
+}
+loc parseUnixPath(str input, loc src=|unknown:///|) = mapPathToLoc(parse(#UnixPath, input, src));
+
+@synopsis{Root is a special case}
+private loc mapPathToLoc((UnixPath) `<Slashes _>`) 
+    = |file:///|;
+
+@synopsis{Absolute: given the drive and relative to its root.}
+private loc mapPathToLoc((UnixPath) `<Slashes _><UnixFilePath path>`) 
+    = appendPath(|file:///|, path);
+
+@synopsis{Relative: relative to the current working directory.}
+private loc mapPathToLoc((UnixPath) `<UnixFilePath path>`) 
+    = appendPath(|cwd:///|, path);
+
+@synopsis{Home relative: relative to the current users home directory}
+private loc mapPathToLoc((UnixPath) `~<Slash _><UnixFilePath path>`) 
+    = appendPath(|home:///|, path);
+
+@synopsis{Home relative: relative to the current users home directory}
+private loc mapPathToLoc((UnixPath) `~`) 
+    = |home:///|;
+
+@synopsis{User relative: relative to any specific user's home directory}
+private loc mapPathToLoc((UnixPath) `~<UserName name><Slash _><UnixFilePath path>`) 
+    = appendPath(|home:///../<name>/|, path);
+
+@synopsis{User relative: relative to any specific user's home directory}
+private loc mapPathToLoc((UnixPath) `~<UserName name>`) 
+    = |home:///../<name>/|;
+
+private loc appendPath(loc root, UnixFilePath path)
+    = (root | it + "<segment>" | segment <- path.segments);
+
+test bool root()
+    = parseUnixPath("/") == |file:///|;
+
+test bool absolutePath()
+    = parseUnixPath("/usr/local/bin")
+    == |file:///usr/local/bin|;
+
+test bool relativePath()
+    = parseUnixPath(".bash_rc")
+    == |cwd:///.bash_rc|;
+
+test bool homePath()
+    = parseUnixPath("~/.bash_profile")
+    == |home:///.bash_profile|;
+
+test bool userPath()
+    = parseUnixPath("~root/.bash_profile")
+    == |home:///../root/.bash_profile|;
```

Comment:
shouldn't this map to `home://root/` ?

Reply:
I tried that but we can not assure cross-platform access to the user directories of a system. It's simply not a portable feature of the Java run-time. So I chose here to transparantly encode the assumption that user directories are siblings. The `home:///` root is resolved using the "user.dir" property that _is_ guaranteed to be portable for the current user. That's the best we can do, or throw out the entire feature.. that has crossed my mind as well.